### PR TITLE
Feature/stacked dictionaries

### PR DESCRIPTION
**Pull request by pgj**
Implement stacking of dictionaries.  It also incorporates some support for checking for potential inconsistencies, such as if two dictionary patch files attempt to override the same message or two dictionaries attempt to implement the same language.

In addition, it introduces Template Haskell to free us from the tedious work of maintaining the translation entries manually.  From now on, it is just enough to add the identifier ("label") of the new message and everything else will be generated automatically (in compile time).

### CODE REVIEW

File: src/Bead/View/DictionaryLoader.hs [hunk 1]
```diff
@@ -67,7 +146,122 @@ loadDictionary path = do
...
-      Just <$> compileExpr "dict :: DictionaryFile" >>= unsafeCoerce
+      let dict = takeBaseName path ++ ".dict"
+      (Just . ((,) path)) <$>
+        compileExpr (dict ++ " :: DictionaryFile") >>= unsafeCoerce
   `catch` (\e -> do
     printf "Could not load dictionary: %s\n" (show (e :: SomeException))
     return Nothing)
+
+#ifdef TEST
+patchDictionariesTests = TS.group "patchDictionaries" $ do
```
Comment (andorp) [inline src/Bead/View/DictionaryLoader.hs]: Please insert patchDictionariesTests into https://github.com/andorp/bead/blob/master/test/Test/Unit/Module.hs

File: src/Bead/View/DictionaryLoader.hs [hunk 2]
```diff
@@ -1,18 +1,27 @@
...
 module Bead.View.DictionaryLoader where
 
 import           Control.Applicative ((<$>))
 import           Control.Arrow ((&&&), (***), (>>>))
 import           Control.Exception
 import           Control.Monad (filterM)
+import           Data.Function
+import           Data.List
 import qualified Data.Map as Map
+#ifdef TEST
```
Comment (andorp) [inline src/Bead/View/DictionaryLoader.hs]: Please unify this #ifdef with the one above.

File: src/Bead/View/DictionaryLoader.hs [hunk 3]
```diff
@@ -67,7 +146,122 @@ loadDictionary path = do
...
+  let aePath = "lang" </> "DictionaryAe.hs"
+  satisfy "Single dictionary"
+    (\case
+      Right m -> fine m "sq" "Shqip"
+      _       -> False)
+    [ (sqPath, DictionaryFile "sq.ico" "sq" "Shqip" []) ]
+    "A single dictionary should work"
+  satisfy "Two different dictionaries"
+    (\case
+      Right m -> fine m "sq" "Shqip" && fine m "ae" "Avesta"
```
Comment (andorp) [inline src/Bead/View/DictionaryLoader.hs]: These are real languages :) Nice.

File: src/Bead/View/DictionaryLoader.hs [hunk 4]
```diff
@@ -67,7 +146,122 @@ loadDictionary path = do
...
+            ]
+  let sqPatches = [ "lang" </> "SqPatch.hs", "lang" </> "SqPatch1.hs", "lang" </> "SqPatch2.hs" ]
+  satisfy "Single dictionary with a patch"
+    (\case
+      Right m -> fine m "sq" "Shqip"
+      _       -> False)
+    [ (sqPath, d)
+    , (sqPatches !! 0, DictionaryPatchFile d [ msg_Login_Username <| "Identifier:" ])
+    ]
+    "A single dictionary with a patch should work"
```
Comment (andorp) [inline src/Bead/View/DictionaryLoader.hs]: Hint could be more specific, like "patch was not applied"

**Comment by pgj** [comment]
Thanks, I have updated the commit.

File: src/Bead/View/DictionaryLoader.hs (hunk 1 above)
Comment (andorp) [inline src/Bead/View/DictionaryLoader.hs]: Resolved in eb0d50a2cbb5fbcfaec5b66aa08608b21bb7b8ad

**Pull request merged**